heat/core/dndarray.py

```diff
@@ -653,6 +653,84 @@ def fill_diagonal(self, value: float) -> DNDarray:
 
         return self
 
+    def __process_key(arr: DNDarray, key: Union[int, Tuple[int, ...], List[int, ...]]) -> Tuple:
+        """
+        Private method for processing keys for indexing. Returns wether advanced indexing is used as well as a processed key and self.
+        A processed key:
+        - doesn't cotain any ellipses or newaxis
+        - all Iterables are converted to ``DNDarrays``
+        - has the same dimensionality as the ``DNDarray`` it indexes
+
+        Parameters
+        ----------
+        key : int, slice, Tuple[int,...], List[int,...]
+            Indices for the tensor.
+        """
+        advanced_indexing = False
+        if isinstance(key, DNDarray):
+            # DNDARRAY CURRENTLY DOES NOT IMPLEMENT the Iterable interface, need to define __iter__()
+            advanced_indexing = True
+            # TODO: check for key.ndim = 0 and treat that as int
+            # TODO: get outshape + outsplit; depends on wether key is bool or int and key.ndim
+        elif isinstance(key, Iterable) and not isinstance(key, tuple):
+            advanced_indexing = True
+            key = factories.array(key)
+            # DOES NOT WORK FOR SEQUENCE OF TENSORS OR DNDARRAYS, works for sequence of ndarrays though
+            # TODO: get outshape + outsplit; depends on wether key is bool or int and key.ndim
+        elif isinstance(key, tuple):
+            key = list(key)
+            for i, k in enumerate(key):
+                if isinstance(k, Iterable) or isinstance(key, DNDarray):
+                    advanced_indexing = True
+                    key[i] = factories.array(key[i])
```

Comment:
`key[i]` might be distributed along `self.split` (or newsplit)

Reply:
What would `newsplit` be?

Reply:
We don't know yet because we need a sanitized key first. I guess what I'm saying is, we need to be agnostic about the `k.split` at this point. How about:

```
                    advanced_indexing = True
                    k_split = getattr(k, "split", None)
                    key[i] = factories.array(key[i], split=k_split)
```

Reply:
Oh no I'm taking this back. tensors and ndarrays have `.split`as well